### ISSUE TITLE
FR: Continuous scroll for CBZ files

### ISSUE THREAD

**Issue by aero31aero** [koreader/koreader#11139]
**Does your feature request involve difficulty completing a task? Please describe.**

Reading webtoons is hard when sometimes dialogs are rendered split up across "pages", and tapping left/right just takes you to the next or previous rendered page.

I'm dealing with webtoons where each image/page in the CBZ file is vertical very tall, containing 2 to 3 panels each. Also, its not necessary that a panel will end where a page ends; panels can seamlessly continue across pages. So, pages don't really mean anything for such content and you want to be able to scroll anywhere in the file.

**Describe the solution you'd like**

Allow vertical scroll like you'd see on a typical webpage.

If not, allow page-overlap for CBZ files (currently, that option is not available for this format).

**Describe alternatives you've considered**

**Additional context**

2 pages of a random CBZ file I have lying around:

<details>
  <summary>Click to expand</summary>
  
![34](https://github.com/koreader/koreader/assets/8033238/f3de0084-5c8f-48de-bb88-9690b275ee63)
![35](https://github.com/koreader/koreader/assets/8033238/f00c16f5-9c38-4e5f-9f14-6952a4545db1)
</details>

**Comment by Frenzie** [comment]
I can't say I understand what you're asking. Not only is it possible, but it's the default?

**Comment by aero31aero** [comment]
Hmm, I've tried my best but haven't managed to get that working. Can you tell how?

My device is a [Kindle 6](https://wiki.mobileread.com/wiki/Kindle_6).

**Comment by hius07** [comment]
You can investigate the raw pictures via the Archive viewer plugin.

**Comment by Frenzie** [comment]
@aero31aero 
> Hmm, I've tried my best but haven't managed to get that working. Can you tell how?

You're saying you can't manage to get it to scroll? Open the bottom menu, fourth icon that looks like text on a page, enable continuous instead of paged if you changed that.

**Comment by hius07** [comment]
And setting Page gap to None gives solid pictures.

**Comment by aero31aero** [comment]
I'm not sure what you mean @hius07, but yeah, I can see that the images/pages are there in the archives. Tapping left/right edges indeed can take me to the next picture, but that's not the behavior I'm looking for. If I set images to fit the screen, then they'll be ant-sized, haha.

@Frenzie, I can confirm I have continuous selected instead of paged, with page-gap as none. I'm not sure how the scroll should behave, am I supposed to swipe up/down?

I've also selected "vertical overlap" as 84% in the third tab, and that allows me to tap left/right to crawl through an individual page, which is fine.

But even with this vertical overlap setting, at the end of a page, a right tap just immediately jumps to the next page instead of showing an inch of the next page and most of the current page.

**Comment by aero31aero** [comment]
FWIW, I can scroll with a mouse-wheel on KOReader on a laptop, but even there, clicking left/right does this "go to previous/next rendered page" thing, not a small scroll, and click+drag vertically does nothing.

So, I guess my question becomes, how do I simulate mouse-scroll events on a Kindle touchscreen?

**Comment by Frenzie** [comment]
> So, I guess my question becomes, how do I simulate mouse-scroll events on a Kindle touchscreen?

You can simply drag it.

(Technically speaking it's the scroll events that are currently implemented as fake panning events; there is actually no such thing as a scroll event in the frontend code. ;-D)

> "go to previous/next rendered page" thing

It's the equivalent of page up/down on a keyboard (which almost never deals in actual pages). So if I understand correctly, when you say "continuous" you mean mini-steps? I'm inclined to think that on E Ink it's much nicer to use next/previous view with occasional drag adjustments if desired instead of constantly triggering full refreshes.

Continuous refers to the next page being rendered partially underneath the end of the current page ("like a webbrowser"), as opposed to actually dealing in pages.

> But even with this vertical overlap setting, at the end of a page, a right tap just immediately jumps to the next page instead of showing an inch of the next page and most of the current page.

If you're seeing the vertical overlap setting then you've changed away from zoom to fit page or content width, which is probably the cause for your issues. It'll even pop up an dialog about it when you try to combine column zoom with continuous or some such but it's not actively disallowed.

Also you might just want a larger page gap in case you set it to none and pages are sticking together too much.

**Comment by aero31aero** [comment]
Hmm, I think I know what's going on now, but I'll still paste the results so you could see for yourself. I made a single 600x6000 image that I then split into some CBZ files (renamed to .zip because of github):

- 600x800 * 7 images - [test-native.zip](https://github.com/koreader/koreader/files/13440368/test-native.zip)
- 600*2000 * 3 images - [test-3.zip](https://github.com/koreader/koreader/files/13440369/test-3.zip)
- 600*6000 * 1 image - [test-1.zip](https://github.com/koreader/koreader/files/13440371/test-1.zip)

Following screenshots display what's happening well:

Each image is the difference of 1 tap.

---

# Single Page Mode

in the third tab, Zoom set to %age and vertical overlap as 60%:

<img src="https://github.com/koreader/koreader/assets/8033238/b6034a09-8eb6-450f-96a3-2aa565a973db" width="150">
<img src="https://github.com/koreader/koreader/assets/8033238/f1a6decd-bbb4-4ff7-a4e0-34ae9898d500" width="150">
<img src="https://github.com/koreader/koreader/assets/8033238/4e02fb8d-ece9-4f22-bd28-0614dc8d7019" width="150">

# Continuous mode

<img src="https://github.com/koreader/koreader/assets/8033238/d54326e4-ec8b-4d00-8004-bb40685ab6fe" witdh="150">
<img src="https://github.com/koreader/koreader/assets/8033238/66835cbf-d275-41e6-830a-a457eda0fa71" witdh="150">

---

So, single page mode is giving me what I want when scrolling through one image, but at page boundaries its going completely to the next page, while continuous mode is by default almost completely going to next page.

> So if I understand correctly, when you say "continuous" you mean mini-steps? I'm inclined to think that on E Ink it's much nicer to use next/previous view with occasional drag adjustments if desired instead of constantly triggering full refreshes.

Yes, and I agree that your occasional drag adjustments sound nice, but I believe I'm being gaslit when you say there is a drag option, you know? I've tried long-press and drag, just regular swipe, 2 finger swipe, etc, nothing does anything.

**Comment by aero31aero** [comment]
So, I essentially want all three cbz files I shared to behave identically, such that I don't even know there are individual images/pages in there. As if its some arbitrarily long 600xinfinite size image, that I'm scrolling through such that any individual scroll/tap takes me no more than halfway through the next chunk.

I believe it's a simple task and I'm missing something, but if not, hopefully a not too complicated feature to implement?

**Comment by Frenzie** [comment]
> So, single page mode is giving me what I want when scrolling through one image, but at page boundaries its going completely to the next page, while continuous mode is by default almost completely going to next page.

Did you try page gap (right under view mode)?

> Any way to debug why drag isn't working for me?

It sounds like you're doing swipes rather than drags, but @poire-z already went into more detail.

**Comment by aero31aero** [comment]
Yeah, I tried page gap. It seems like a mere cosmetic thingy, inserting a grey border between pages. It doesn't seem to modify the scroll behavior at all.

I did manage to hack something together which I'm now trying to somehow build for my Kindle but failing (ugh!), so I cleaned it up a bit and made a PR in the meantime..

**Comment by NiLuJe** [comment]
> As for drags, I finally managed to drag things and I see what you mean; its essentially what I want, but at the same time, maybe the drag detection code doesn't work on Kindles well? I am only able to trigger it once every 4 or so attempts whenever I'm trying it.

You might simply need to tweak the swipe threshold in the Gestures setup menu?

(IIRC, I pretty much halve or even quarter *all* the time-related thresholds over here).
